File: Data Structures & Algorithms/burst-balloons/submission-0.java
class Solution {
    int brust(int[] arr, int l, int r , int[][]dp) {
        if (l > r) {
            return 0;
        }
        
        if (dp[l][r] != -1) {
            return dp[l][r];
        }
        // 1 2 4 3 5
        for (int i = l; i <= r; i++) {
            int coins = arr[l-1] * arr[i] * arr[r+1];
            coins += brust(arr, l, i - 1, dp);
            coins += brust(arr, i + 1, r, dp);
            dp[l][r] = Math.max(dp[l][r], coins);
        }
        return dp[l][r];
    }
    public int maxCoins(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        
        int n = nums.length;
        int[] newNums = new int[n + 2];
        newNums[0] = newNums[n + 1] = 1;
        for (int i = 0; i < n; i++) {
            newNums[i + 1] = nums[i];
        }

        int[][] dp = new int[n + 2][n + 2];
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= n; j++) {
                dp[i][j] = -1;
            }
        }

        return brust(newNums, 1, newNums.length - 2, dp);
    }
}

File: Data Structures & Algorithms/longest-common-subsequence/submission-0.cs
public class Solution {
    private int[,] dp;

    private int LCS(int i, int j, string s, string t) {
        if (i == s.Length || j == t.Length) {
            return 0;
        }

        if (dp[i,j] != -1) {
            return dp[i,j];
        }

        if (s[i] == t[j]) {
            dp[i,j] = 1 + LCS(i + 1, j + 1, s, t);
        } else {
            dp[i,j] = Math.Max(LCS(i + 1, j, s, t), LCS(i, j + 1, s, t));
        }
        return dp[i,j];
    }

    public int LongestCommonSubsequence(string text1, string text2) {
        dp = new int [text1.Length, text2.Length];

        for (int i = 0; i < text1.Length; i++) {
            for (int j = 0; j < text2.Length; j++) {
                dp[i,j] = -1;
            }
        }
        return LCS(0, 0, text1, text2);
    }
}
